network/p2p/peerstore/peerstore.go

```diff
@@ -148,26 +152,23 @@ func (ps *PeerStore) GetConnectionWaitTime(addrOrPeerID string) (bool, time.Dura
 	}
 
 	// Remove the expired elements from e.data[addr].recentConnectionTimes
-	ps.popNElements(numElmtsToRemove, peerID)
+	ad.recentConnectionTimes = ad.recentConnectionTimes[numElmtsToRemove:]
+
 	// If there are max number of connections within the time window, wait
-	metadata, _ = ps.Get(peerID, addressDataKey)
-	ad, ok = metadata.(addressData)
-	if !ok {
-		return false, 0 /* not used */, curTime /* not used */
-	}
-	numElts := len(ad.recentConnectionTimes)
-	if uint(numElts) >= ps.connectionsRateLimitingCount {
-		return true, /* true */
-			ps.connectionsRateLimitingWindow - timeSince, curTime /* not used */
+	remainingLength := originalLen - numElmtsToRemove
+	var waitTime time.Duration
+	if uint(remainingLength) >= ps.connectionsRateLimitingCount {
+		waitTime = ps.connectionsRateLimitingWindow - timeSince
+	} else {
+		// Else, there is space in connectionsRateLimitingCount. The
+		// connection request of the caller will proceed
+		// Append the provisional time for the next connection request
+		ad.recentConnectionTimes = append(ad.recentConnectionTimes, curTime)
 	}
+	ad.mu.Unlock()
```

Comment:
I don't really follow why we change an `ad` under a  lock, but we don't protect the `ps` itself across `Get/Put` pairs.  If this and another manipulator of the same ad is called in two threads, there can be a `Get` followed by manipulation (which might change the ad in two different ways), the `Puts` happen in _some_ order, so one of them is lost.

Reply:
I agree, peerstore needs to be re-written to have a lock like phonebook, and not per-entry locks that do not serve synchronization well.
Rationale: `ad` entry (and so that `ad.mu` is appears to be created once when a peerID is seen a first time. `ad.mu.Lock` + `ps.Put` + `ad.mu.Unlock` sequence still not safe b/c a thread get be getting a stale `ad` when calling `ad.Get` without a lock.

Reply:
done, added mutex to `PeerStore` itself similar to `Phonebook`